### ISSUE TITLE
Extraction of gzipped sitemap fails in Scrapy 0.24.4

### ISSUE THREAD

**Issue by mercutio79** [scrapy/scrapy#951]
retrieving a gzipped sitemap xml (tested on amazon.de) fails.

Reproduce with :
modify /utils/gz.py gunzip method to write the incoming data to a file.

gunzip the file on the command line.

the unzipped file contains garbled content

gunzip that file with garbled content a second time and get the correct content

-> I suspect that the content coming from the target server is already gzip compressed and scrapy has a bug that causes the gzip decompression to not work properly, resulting in a double compressed file arriving at the  /utils/gz.py gunzip method

**Comment by dszmaj** [comment]
i've encountered the same problem with redirect downloads, without redirect everything is fine, it looks that only after redirect, content gets double gzipped

workaraund from @mercutio79 works perfectly with this

**Comment by juraseg** [comment]
Compression middleware behaviour on gzipped sitemaps still seems to be wrong. 

In case of amazon it does not work anymore, because it sets header "Content-Type" to "application/octet-stream" instead of "application/x-gzip". So this for this issue - [https://github.com/scrapy/scrapy/issues/193](Issue #193) - does not work anymore. It still tries to decompress the file two times - once in HTTPCompressionMiddleware and once in Sitemap spider.

I've encountered opposite issue on other site. Sitemap file is also gzipped and has extension ".gz", but when site gets header "Accept-Encoding"="x-gzip,gzip,deflate" (which is added by compression middleware) it compresses the file once more. Response contents in this case is twice compressed. This is of course an error on site hosters side, but in my opinion Scrapy should account for this anyway - I imagine there are a lot of such sites in the web.


To get rid of all of these issues I suggest to move the whole compression behavior to HTTPCompressionMiddleware, and it should perform decompressing in presence of only "Content-Encoding" header (disregard check for "Content-Type") and do decompression in a loop until gets an error (easier to ask forgiveness than permission) - gzip module will fail earlier of first read so should not add much overhead. Another option is to check gzip file format, as suggested by @dangra in https://github.com/scrapy/scrapy/issues/193](Issue #193). This would be a more complex solution, but potentially will perform better.

Also, to account for cases when HTTPCompressionMiddleware is disabled, Sitemap can still try to decompress response content if extension is ".gz".

**Comment by redapple** [comment]
Fixed in scrapy 1.1.1 (https://github.com/scrapy/scrapy/pull/2065):

Example spider for amazon.fr sitemap:

```
$ cat gzippedsitemaps/spiders/amz.py
from scrapy.spiders import SitemapSpider

class MySpider(SitemapSpider):
    name = "amz"
    sitemap_urls = ['https://www.amazon.fr/sitemaps.2307ea63773dfee.SitemapIndex_0.xml.gz']

    def parse(self, response):
        pass
```

scrapy 1.1, Python 2.7

```
$ scrapy version -v
Scrapy    : 1.1.2
lxml      : 3.6.4.0
libxml2   : 2.9.4
Twisted   : 16.3.2
Python    : 2.7.11+ (default, Apr 17 2016, 14:00:29) - [GCC 5.3.1 20160413]
pyOpenSSL : 16.0.0 (OpenSSL 1.0.2g-fips  1 Mar 2016)
Platform  : Linux-4.4.0-36-generic-x86_64-with-Ubuntu-16.04-xenial

$ scrapy crawl amz
2016-09-14 15:21:46 [scrapy] INFO: Scrapy 1.1.2 started (bot: gzippedsitemaps)
2016-09-14 15:21:46 [scrapy] INFO: Overridden settings: {'NEWSPIDER_MODULE': 'gzippedsitemaps.spiders', 'SPIDER_MODULES': ['gzippedsitemaps.spiders'], 'ROBOTSTXT_OBEY': True, 'BOT_NAME': 'gzippedsitemaps'}
(...)
2016-09-14 15:21:46 [scrapy] INFO: Spider opened
2016-09-14 15:21:46 [scrapy] INFO: Crawled 0 pages (at 0 pages/min), scraped 0 items (at 0 items/min)
2016-09-14 15:21:46 [scrapy] DEBUG: Telnet console listening on 127.0.0.1:6023
2016-09-14 15:21:46 [scrapy] DEBUG: Crawled (200) <GET https://www.amazon.fr/robots.txt> (referer: None)
2016-09-14 15:21:47 [scrapy] DEBUG: Crawled (200) <GET https://www.amazon.fr/sitemaps.2307ea63773dfee.SitemapIndex_0.xml.gz> (referer: None)
2016-09-14 15:21:53 [scrapy] DEBUG: Crawled (200) <GET https://www.amazon.fr/sitemaps.2307ea63773dfee.sitemap_sitemap_fr_dirty_detail_tsv_0_25_6_5.xml.gz> (referer: https://www.amazon.fr/sitemaps.2307ea63773dfee.SitemapIndex_0.xml.gz)
2016-09-14 15:21:54 [scrapy] DEBUG: Crawled (200) <GET https://www.amazon.fr/sitemaps.2307ea63773dfee.sitemap_sitemap_fr_dirty_detail_tsv_0_25_5_4.xml.gz> (referer: https://www.amazon.fr/sitemaps.2307ea63773dfee.SitemapIndex_0.xml.gz)
2016-09-14 15:21:54 [scrapy] DEBUG: Crawled (200) <GET https://www.amazon.fr/sitemaps.2307ea63773dfee.sitemap_sitemap_fr_dirty_detail_tsv_0_25_9_8.xml.gz> (referer: https://www.amazon.fr/sitemaps.2307ea63773dfee.SitemapIndex_0.xml.gz)
(...)
```

scrapy 1.1, Python 3.5
```
$ scrapy version -v
Scrapy    : 1.1.2
lxml      : 3.6.4.0
libxml2   : 2.9.4
Twisted   : 16.4.0
Python    : 3.5.1+ (default, Mar 30 2016, 22:46:26) - [GCC 5.3.1 20160330]
pyOpenSSL : 16.1.0 (OpenSSL 1.0.2g-fips  1 Mar 2016)
Platform  : Linux-4.4.0-36-generic-x86_64-with-Ubuntu-16.04-xenial

$ scrapy crawl amz
2016-09-14 15:16:52 [scrapy] INFO: Scrapy 1.1.2 started (bot: gzippedsitemaps)
(...)
2016-09-14 15:16:52 [scrapy] INFO: Spider opened
2016-09-14 15:16:52 [scrapy] INFO: Crawled 0 pages (at 0 pages/min), scraped 0 items (at 0 items/min)
2016-09-14 15:16:52 [scrapy] DEBUG: Telnet console listening on 127.0.0.1:6023
2016-09-14 15:16:53 [scrapy] DEBUG: Crawled (200) <GET https://www.amazon.fr/robots.txt> (referer: None)
2016-09-14 15:16:53 [scrapy] DEBUG: Crawled (200) <GET https://www.amazon.fr/sitemaps.2307ea63773dfee.SitemapIndex_0.xml.gz> (referer: None)
2016-09-14 15:16:58 [scrapy] DEBUG: Crawled (200) <GET https://www.amazon.fr/sitemaps.2307ea63773dfee.sitemap_sitemap_fr_dirty_detail_tsv_0_25_20_19.xml.gz> (referer: https://www.amazon.fr/sitemaps.2307ea63773dfee.SitemapIndex_0.xml.gz)
(...)
```

scrapy 1.0.6, Python 2.7

```
$ scrapy version -v
Scrapy    : 1.0.6
lxml      : 3.6.1.0
libxml2   : 2.9.3
Twisted   : 16.3.0
Python    : 2.7.9 (default, Apr  2 2015, 15:33:21) - [GCC 4.9.2]
pyOpenSSL : 16.0.0 (OpenSSL 1.0.2g-fips  1 Mar 2016)
Platform  : Linux-4.4.0-36-generic-x86_64-with-Ubuntu-16.04-xenial

$ scrapy crawl amz
2016-09-14 15:20:30 [scrapy] INFO: Scrapy 1.0.6 started (bot: gzippedsitemaps)
2016-09-14 15:20:30 [scrapy] INFO: Optional features available: ssl, http11
2016-09-14 15:20:30 [scrapy] INFO: Overridden settings: {'NEWSPIDER_MODULE': 'gzippedsitemaps.spiders', 'SPIDER_MODULES': ['gzippedsitemaps.spiders'], 'ROBOTSTXT_OBEY': True, 'BOT_NAME': 'gzippedsitemaps'}
2016-09-14 15:20:30 [scrapy] INFO: Enabled extensions: CloseSpider, TelnetConsole, LogStats, CoreStats, SpiderState
2016-09-14 15:20:30 [scrapy] INFO: Enabled downloader middlewares: RobotsTxtMiddleware, HttpAuthMiddleware, DownloadTimeoutMiddleware, UserAgentMiddleware, RetryMiddleware, DefaultHeadersMiddleware, MetaRefreshMiddleware, HttpCompressionMiddleware, RedirectMiddleware, CookiesMiddleware, ChunkedTransferMiddleware, DownloaderStats
2016-09-14 15:20:30 [scrapy] INFO: Enabled spider middlewares: HttpErrorMiddleware, OffsiteMiddleware, RefererMiddleware, UrlLengthMiddleware, DepthMiddleware
2016-09-14 15:20:30 [scrapy] INFO: Enabled item pipelines: 
2016-09-14 15:20:30 [scrapy] INFO: Spider opened
2016-09-14 15:20:30 [scrapy] INFO: Crawled 0 pages (at 0 pages/min), scraped 0 items (at 0 items/min)
2016-09-14 15:20:30 [scrapy] DEBUG: Telnet console listening on 127.0.0.1:6023
2016-09-14 15:20:30 [scrapy] DEBUG: Crawled (200) <GET https://www.amazon.fr/robots.txt> (referer: None)
2016-09-14 15:20:30 [scrapy] DEBUG: Crawled (200) <GET https://www.amazon.fr/sitemaps.2307ea63773dfee.SitemapIndex_0.xml.gz> (referer: None)
2016-09-14 15:20:37 [scrapy] DEBUG: Crawled (200) <GET https://www.amazon.fr/sitemaps.2307ea63773dfee.sitemap_sitemap_fr_dirty_detail_tsv_1_26_11_10.xml.gz> (referer: https://www.amazon.fr/sitemaps.2307ea63773dfee.SitemapIndex_0.xml.gz)
2016-09-14 15:20:37 [scrapy] ERROR: Spider error processing <GET https://www.amazon.fr/sitemaps.2307ea63773dfee.sitemap_sitemap_fr_dirty_detail_tsv_1_26_11_10.xml.gz> (referer: https://www.amazon.fr/sitemaps.2307ea63773dfee.SitemapIndex_0.xml.gz)
Traceback (most recent call last):
  File "/home/paul/.virtualenvs/scrapy10/local/lib/python2.7/site-packages/scrapy/utils/defer.py", line 102, in iter_errback
    yield next(it)
  File "/home/paul/.virtualenvs/scrapy10/local/lib/python2.7/site-packages/scrapy/spidermiddlewares/offsite.py", line 28, in process_spider_output
    for x in result:
  File "/home/paul/.virtualenvs/scrapy10/local/lib/python2.7/site-packages/scrapy/spidermiddlewares/referer.py", line 22, in <genexpr>
    return (_set_referer(r) for r in result or ())
  File "/home/paul/.virtualenvs/scrapy10/local/lib/python2.7/site-packages/scrapy/spidermiddlewares/urllength.py", line 37, in <genexpr>
    return (r for r in result or () if _filter(r))
  File "/home/paul/.virtualenvs/scrapy10/local/lib/python2.7/site-packages/scrapy/spidermiddlewares/depth.py", line 54, in <genexpr>
    return (r for r in result or () if _filter(r))
  File "/home/paul/.virtualenvs/scrapy10/local/lib/python2.7/site-packages/scrapy/spiders/sitemap.py", line 36, in _parse_sitemap
    body = self._get_sitemap_body(response)
  File "/home/paul/.virtualenvs/scrapy10/local/lib/python2.7/site-packages/scrapy/spiders/sitemap.py", line 65, in _get_sitemap_body
    return gunzip(response.body)
  File "/home/paul/.virtualenvs/scrapy10/local/lib/python2.7/site-packages/scrapy/utils/gz.py", line 20, in gunzip
    chunk = f.read(8196)
  File "/usr/lib/python2.7/gzip.py", line 268, in read
    self._read(readsize)
  File "/usr/lib/python2.7/gzip.py", line 303, in _read
    self._read_gzip_header()
  File "/usr/lib/python2.7/gzip.py", line 197, in _read_gzip_header
    raise IOError, 'Not a gzipped file'
IOError: Not a gzipped file
2016-09-14 15:20:37 [scrapy] DEBUG: Crawled (200) <GET https://www.amazon.fr/sitemaps.2307ea63773dfee.sitemap_sitemap_fr_dirty_detail_tsv_1_26_12_11.xml.gz> (referer: https://www.amazon.fr/sitemaps.2307ea63773dfee.SitemapIndex_0.xml.gz)
2016-09-14 15:20:37 [scrapy] ERROR: Spider error processing <GET https://www.amazon.fr/sitemaps.2307ea63773dfee.sitemap_sitemap_fr_dirty_detail_tsv_1_26_12_11.xml.gz> (referer: https://www.amazon.fr/sitemaps.2307ea63773dfee.SitemapIndex_0.xml.gz)
Traceback (most recent call last):
  File "/home/paul/.virtualenvs/scrapy10/local/lib/python2.7/site-packages/scrapy/utils/defer.py", line 102, in iter_errback
    yield next(it)
  File "/home/paul/.virtualenvs/scrapy10/local/lib/python2.7/site-packages/scrapy/spidermiddlewares/offsite.py", line 28, in process_spider_output
    for x in result:
  File "/home/paul/.virtualenvs/scrapy10/local/lib/python2.7/site-packages/scrapy/spidermiddlewares/referer.py", line 22, in <genexpr>
    return (_set_referer(r) for r in result or ())
  File "/home/paul/.virtualenvs/scrapy10/local/lib/python2.7/site-packages/scrapy/spidermiddlewares/urllength.py", line 37, in <genexpr>
    return (r for r in result or () if _filter(r))
  File "/home/paul/.virtualenvs/scrapy10/local/lib/python2.7/site-packages/scrapy/spidermiddlewares/depth.py", line 54, in <genexpr>
    return (r for r in result or () if _filter(r))
  File "/home/paul/.virtualenvs/scrapy10/local/lib/python2.7/site-packages/scrapy/spiders/sitemap.py", line 36, in _parse_sitemap
    body = self._get_sitemap_body(response)
  File "/home/paul/.virtualenvs/scrapy10/local/lib/python2.7/site-packages/scrapy/spiders/sitemap.py", line 65, in _get_sitemap_body
    return gunzip(response.body)
  File "/home/paul/.virtualenvs/scrapy10/local/lib/python2.7/site-packages/scrapy/utils/gz.py", line 20, in gunzip
    chunk = f.read(8196)
  File "/usr/lib/python2.7/gzip.py", line 268, in read
    self._read(readsize)
  File "/usr/lib/python2.7/gzip.py", line 303, in _read
    self._read_gzip_header()
  File "/usr/lib/python2.7/gzip.py", line 197, in _read_gzip_header
    raise IOError, 'Not a gzipped file'
IOError: Not a gzipped file
(...)
```
